docs/se/guides/migration_4x.adoc

```diff
@@ -0,0 +1,216 @@
+
+In Helidon 4 the routing is configured the following way:
+
+[source, java]
+.Start Helidon SE 4.x Server
+----
+static void routing(HttpRouting.Builder routing) {
+    Config config = Config.global();
+
+    routing.register("/greet", new GreetService())  <1>
+            .addFeature(OpenApiFeature.create(config.get("openapi")))   <2>
+            .addFeature(ObserveFeature.create(config.get("observe")));
+}
+----
+<1> Register Greeting service as in previous versions of Helidon.
+<2> Add `Feature` for Observability and OpenAPI.
+
+
+`Feature` encapsulates a set of endpoints, services and/or filters. It is similar to `HttpService` but gives more freedom in setup. Main difference is that a feature can add `Filters` and it cannot be registered on a path. Features are not registered immediately - each feature can define a `Weight` or implement `Weighted` to order features according to their weight. Higher weighted features are registered first. This is to allow ordering of features in a meaningful way (e.g. Context should be first, Tracing second, Security third etc.).
+
+
+== Services
+
+There are also significant changes in Helidon `Service`.
+
+In prior versions a service look this way:
+
+[source, java]
+.Helidon SE 3.x Service
+----
+public class GreetService implements Service {
+
+    @Override
+    public void update(Routing.Rules rules) {   <1>
+        rules
+            .get("/", this::getDefaultMessageHandler)
+            .get("/{name}", this::getMessageHandler)
+            .put("/greeting", this::updateGreetingHandler);
+    }
+
+    private void getDefaultMessageHandler(ServerRequest request, ServerResponse response) {   <2>
+        sendResponse(response, "World");
+    }
+
+    // other methods omitted
+}
+----
+<1> Use `update()` method to set up routing.
+<2> Handle a `Request` and return a `Responce`.
+
+In Helidon 4 the same service:
+
+[source, java]
+.Helidon SE 4.x Service
+----
+public class GreetService implements HttpService {  <1>
+
+    @Override
+    public void routing(HttpRules rules) {  <2>
+        rules.get("/", this::getDefaultMessageHandler)
+             .get("/{name}", this::getMessageHandler)
+             .put("/greeting", this::updateGreetingHandler);
+    }
+
+
+    private void getDefaultMessageHandler(ServerRequest request, ServerResponse response) { <3>
+        sendResponse(response, "World");
+    }
+
+    // other methods omitted
+}
+----
+<1> Implement `HttpService` for the `GreetingService`.
+<2> Use `routing(HttpRules rules)` to set up routing.
+<3> Handle a `Request` and return a `Responce`.
+
+Helidon 4 introduced `HttpService` that should be implemented in order to process HTTP requests. To set up routing, the method `routing(HttpRules rules)` should now be used. It receives `HttpRules` object with routes description.
+
+These changes make Helidon 4 incompatible with previous versions.
+
+Learn more about `HttpService` and `Routing` at xref:../webserver.adoc[Helidon SE WebServer]
+
```

Comment:
Maybe add a section on Global Config (see SE Config docs) and a section on Observability including the difference in endpoints.

Reply:
will be done